index.bs

```diff
@@ -176,3 +176,86 @@ Acknowledgements {#ack}
 
 Cure53's [[DOMPURIFY]] is a clear inspiration for the API this document
 describes, as is Internet Explorer's {{window.toStaticHTML()}}.
+
+
+# Appendix A: Integration with Trusted Types {#trustedtypes}
+
+Issue(WICG/sanitizer-api#20): The best way to integrate Trusted Types is still under discussion.
+
+Issue: The two methods below
+  (<a href="#trustedtypes-sanitize-to">`sanitizeToTrustedHTML`</a> and
+   <a href="#trustedtypes-createHTML">createHTML</a>)
+  are redundant. It is unclear whether both are needed.
+
+[Trusted Types](https://w3c.github.io/webappsec-trusted-types/dist/spec/) is
+a proposed feature to allow locking down the DOM to prevent cross-site
+scripting attacks. Trusted Types is complementary to the mechanisms proposed
+here. User Agents that support both, this Sanitizer API and Trusted Types,
+require integration between these features.
+
+## `Sanitizer.sanitize` sink. {#trustedtypes-sink}
+
+In Trusted Types parlance, the Sanitizer.sanitize method is an HTML sink, and
+thus needs to be guarded by the usual Trusted Types mechanisms.
+
+<pre>
+  typedef (DOMString or DocumentFragment or Document) SanitizerInput;
+  <ins>typedef ([StringContext=HTML]DOMString or TrustedHTML or DocumentFragment or Document) TrustedTypeSanitizerInput;</ins>
+
+  // In interface Sanitizer:
+  DocumentFragment sanitize(<ins>TrustedTypes</ins>SanitizerInput input);
+</pre>
+
+## `Sanitizer.sanitizeToTrustedHTML`. {#trustedtypes-sanitize-to}
+
+To allow easy combination of Sanitizer and Trusted Types, the `Sanitizer`
+interface gets a new method to create a Trusted Type directly:
+
+<pre>
+  // In interface Sanitizer:
+  <ins>TrustedHTML sanitizeToTrustedHTML(TrustedTypesSanitizerInput input);</ins>
+</pre>
+
+Since this is - naturally - a new sink, it requires a Trusted Types check.
+But it clearly makes the sense to use this as a way to create Trusted Types
+from untrusted input. To facilitate this, the
+<a href="https://w3c.github.io/webappsec-trusted-types/dist/spec/#trusted-types-csp-directive">`trusted-types` CSP directive</a>
+gets a new keyword `'sanitizer-api'`. If set, the `sanitizeToTrustedHTML` method
```

Comment:
You might want to move this to require-trusted-types-for keyword, e.g.

`require-trusted-types-for 'script' 'allow-sanitizer-api'`

This is the directive that controls sinks. `trusted-types` controls policy creation.

Reply:
Hm. I had seen this as effectively creating implicit policies, hence `trusted-types` rather than `require-trusted-types-for`. I can see how it could also be a seperate class for sinks.

I do think seeing this as a sink class is a little awkward, since having a sanitizer create `TrustedHTML` instances falls into the domain of `script`.

(I don't feel super strongly, though.)

Reply:
I see. That makes sense. The keyword should only control the `Sanitizer.sanitize*` functions, the SanitizerConfig usage in the default policy can be unrestricted, as whoever controls the default policy can just as well _not_ sanitize at all, so we're not increasing the risk here.

Reply:
Coming to think of it, most of the times one does not need `trusted-types` directive at all, and this would force those authors to add `trusted-types * 'allow-duplicates' 'sanitizer-api'`, which might be an overkill. 

Perhaps we don't actually need the keyword at all? We're guarding on 'script' sinks for the purpose of DOM XSS. If the sanitizer is guaranteed never to produce a scripty output, it's equivalent to calling native DOM APIs (`createElement`, `setAttribute`, `dataset` and such) which we don't guard on. In fact sanitizer is more strict, as also "javascript:URLs" are rejected. Potentially dropping the keyword  however requires that the baseline of "sanitizer API output does not contain JS" is preserved.

// cc @xtofian

Reply:
I do not like the idea of `'sanitizer-api'` keyword. I would like `sanitizeToTrustedHTML` to only return safe HTML (i.e. if `SanitizerConfig` is unsafe, `sanitizeToTrustedHTML` method will not be exposed or throw an error with `trustedTypes.emptyHTML` as a return value).

I believe that this will benefit a lot of JS frameworks where they don't have control over HTML or response header. If we require `'sanitizer-api'` keyword, then usage of Sanitizer API in frameworks will break all websites that uses the library with Trusted Types.

Reply:
The invariant to preserve is that (under `require-trusted-types-for 'script'` for all involved documents) the ability to execute Javascript from a string needs a policy (creation of which might be controlled by `trusted-types`). 

If the sanitizer output - under any configuration - can also be generated with regular non-TT-guarded DOM APIs, it's fine to drop the keyword, as the sanitizer does not give any concerning capability - and the sanitizer as such can create `TrustedHTML` instances. It's just like `trustedTypes.emptyHTML - you don't need a policy for it, as it can't cause XSS. I think the current intention for the API is to be inherently secure like this. 

If this changes and it's possible to output XSSy trees (e.g. some configurations would allow `<iframe srcdoc>`), the integration needs a policy or other opt-in mechanism.

Reply:
I don't think we'll allow the sanitizer to produce what you call "XSSy trees". The built-in allow list ought to only be expandable by custom elements (and, if feasible, only those custom elements that _inherit_ from an allowed element)

Reply:
I don't think we'll allow the sanitizer to produce what you call "XSSy trees". The built-in allow list ought to only be expandable by custom elements (and, if feasible, only those custom elements that _inherit_ from an allowed element)

Reply:
Double like :)

Reply:
Hmm. So: They way we're currently spec-ing (and implementing) Sanitizer the user cannot loosen the built-in defaults, and furthermore the defaults - while not defined yet - should cover all the XSS sources that Trusted Types covers as well. (I think everyone agrees on this.)

The way I look at TT is that is gives you control over the string-to-DOM transition. In that case, an on/off switch in CSP for Santizer-to-TrustedHTML would make a lot of sense. I like koto's definition of "the invariant to preserve is that [...]  the ability to execute Javascript from a string needs a policy". If we agree on that, the on/off switch is not required. In fact, it wouldn't make sense.

Do we agree on that? (If so, I'll just drop the CSP bits. I think so far I'm the only hold-out.)

Also: If we're there: Do we need any integration with TT at all?

We can add the sanitizeToTrustedHTML method, but if santizer (to DocumentFragment) is not-TT checked because it's inherently XSS-safe, then this doesn't buy us much. In any case, I think we can drop both the CSP bits; the TT check; and the policy-from-SanitzerConfig integration. (The last bit was meant to make it easier to use Sanitizer within TT, but if any Sanitizer can generate TT instances directly, that's just an extra indirection.)

Reply:
> The way I look at TT is that is gives you control over the string-to-DOM transition

Not really. It guards string => XSS sinks. You can create DOM from strings without TT policies already, that's expected. What you cannot create is XSSy DOM. The only reason why we're guarding HTML sinks is because the resulting tree could execute JS.

> We can add the sanitizeToTrustedHTML method, but if santizer (to DocumentFragment) is not-TT checked because it's inherently XSS-safe, then this doesn't buy us much. 

It actually does buy as a lot. It's a terrific way of obtaining safe serializations to pass to HTML sinks, even under TT restrictions. TT + a non-policy way of creating safe HTML allows you to have real-world  applications that work under `require-trusted-types-for 'script'; trusted-types 'none'`, in other words - there is NO code to review, and no JS to inject to every page for security. Everything is guaranteed to be XSS free (modulo some cross-document concerns). Without `sanitizeToTrustedHTML` you can only have that *if* you're not using `innerHTML`, so practically never.

> Drop policy-from-SanitzerConfig integration. 

It potentially allows for optimization, as you'd only have a single, non-exposed instance used internally per-policy.  But I agree it's not a huge advantage.